FIX#280363

sidewayss opened the pull request:
This is a multi-issue [EPIC] set of fixes that needs UI review and approval.  See the attached docs  in PDF format for details and justifications.

I have placed documentary comments (to be removed later) throughout the code starting with: ///!!!
On Windows 10 at least, both Qt Creator and VS2017 style /// differently from //, so these comments are highlighted inside those editors.

[MS_PageSettings.pdf](https://github.com/musescore/MuseScore/files/2983139/MS_PageSettings.pdf)

sidewayss commented on mscore/prefsdialog.cpp:
```diff
@@ -1385,10 +1411,13 @@ void PreferenceDialog::printShortcutsClicked()
-      qreal pageH = s.value(Sid::pageHeight).toReal();
-      printer.setPaperSize(QSizeF(pageW, pageH), QPrinter::Inch);
+      ///!!!printer.setPaperSize() is no longer in the Qt docs for QPrinter
+      ///!!!is it obsolete??? I am switching to the documented setPageSize() function
```
setPageLayout() not setPageSize()

sidewayss commented:
Given that this PR will not be included in the 3.0 release, this build might well be the final build for this PR.  The compatibility will likely need to be maintained through the lifetime of v3.  But that is up to you all in the control tower at MuseScore Central.  This build can be integrated with no compatibility issues, AFAIK.  But it does have some extra code to handle file format version 3.01 versus 3.02.

Jojo-Schmitz commented:
cgheck https://travis-ci.org/musescore/MuseScore/jobs/473330365#L5003-L5040 it seems the mtest code doesn't see `Ms::MScore::sizesMetric` and friends

sidewayss commented:
Interesting.  Thanks.  It works in VS2017...  I'll see what I can to do provide the necessary visibility and force push an update.

Jojo-Schmitz commented:
that doesn't build the mtests

sidewayss commented:
That's a bummer!  So I have to build it in Qt?  I just searched the solution in VS2017 and it finds nothing for "mtest".  Why is this configured this way?  I seem to remember that Qt built the mtest sub-project, but I haven't built with Qt in a while.

Jojo-Schmitz commented:
No, you just need to tell the mtest code where to find the object containing those definitions and linke it.

sidewayss commented:
But I need to open the files manually (or via search across files outside the project) in VS2017.
And I won't be able to compile it and test it locally.  So Travis is my compiler until I resolve this and any further issues that arise?
I just added a `#include "mscore.h` in \MuseScore\mtest\zerberus\envelopes\tst_sfzenvelopes.cpp.  Is that what you are suggesting?

Jojo-Schmitz commented:
Not sure this is sufficient, whether that particular test is the only one failing to build. More loike needed to add this to mtest/mtest.cpp

sidewayss commented:
OK, it's mscore.h, which is included in soooo many places I would never have thunk it could be an issue.
I just committed the single change so I'll see how it goes.  I think I botched the commit message too...
Also, am I allowed to have multiple commits on this PR branch?

Jojo-Schmitz commented:
You may want to join the developers' chat at https://t.me/musescoreeditorchat

sidewayss commented:
I just joined the chat - did you mean I should ask the same questions there, or that you had a conversation in mind?  I have saved links with instructions for squashing commits, so I'll dig those up.  I have done it previously, but not recently.

sidewayss commented:
At least there are no more link errors, and I've removed the includes of mscore/preferences.h in the libmscore source files.
Now it is the mtests themselves that are failing, after they compile and link.  I wonder if the file format changes interfere with the tests.  They shouldn't interfere with opening .mscz files.  I don't know enough about the tests themselves to opine.

mattmcclinch commented on libmscore/style.cpp:
```diff
@@ -79,6 +79,7 @@ static const StyleType styleTypes[] { // page sizes and margins in points w/deci
       { Sid::pageUnits,               "pageUnits",                 0            }, // QPageLayout::Unit
       { Sid::pageOrientation,         "pageOrientation",           0            }, // QPageLayout::Orientation 
       { Sid::marginOddLeft,           "marginOddLeft",            10.0/INCH*PPI },
+      { Sid::pageOddRightMargin,      "marginOddLeft",            10.0/INCH*PPI },
```
The second column should be "pageOddRightMargin".

sidewayss replied:
Thanks again.  I did the minimum to get it past Travis, which is my current focus.  I figured if it compiled I was OK.  But as I think about it now, while Travis is running, the mtests might fail, if they verify values.  I'll fix this and wait for this Travis to finish before committing and pushing again.
Regardless, this would be an issue at run-time in MuseScore because I failed to edit the string when I reinserted pageOddRightMargin.

mattmcclinch commented on mscore/preferences.cpp:
```diff
@@ -194,6 +196,9 @@ void Preferences::init(bool storeInMemoryOnly)
 
+      QLocale* local = new QLocale();
+      set(PREF_APP_PAGE_UNITS_VALUE, local->measurementSystem() == QLocale::MetricSystem 
+                                   ? int(QPageSize::Millimeter) :  int(QPageSize::Inch));
```
This needs to come after
```
_initialized = true;
```

sidewayss replied:
Interesting.  I had not seen any issues while debugging and testing.  I just moved it down, but I'm curious what problem this would cause as is.

mattmcclinch replied:
Otherwise it fails the assertion in Preferences::settings().

mattmcclinch commented:
Once you are certain that your code is producing the correct MSCX, you can modify the -ref.mscx files to match the generated results. If you see an unexpected difference, that is an indication that something is going wrong.

sidewayss commented:
And then Travis uses my branch's copies of those files, thus they are part of the PR.  That makes sense, and I'm modifying the score files, not the .cpp files.  I have to download the log and filter/sort/organize the list of affected files, which seems to be quite long.  I'll test on the current master opening the new file format prior to committing.

mattmcclinch commented on mscore/pagesettings.cpp:
```diff
@@ -109,158 +111,211 @@ void PageSettings::setScore(Score* s)
+      case PAPER_TYPE_OTHER:
+            _sizes = &MScore::sizesOther;
+            break;
+      }
```
This brace and the entire body of the switch should be indented one more level.

sidewayss replied:
Fixed.  I was unsure about the switch indentation.  A JavaScript editor's auto-indent started me doing it that way.  I think it was Qt Creator with JavaScript, actually.

mattmcclinch commented on mscore/pagesettings.cpp:
```diff
@@ -109,158 +111,211 @@ void PageSettings::setScore(Score* s)
+            QRectF rect = qps->rect(unit);
+            updateWidthHeight(rect);
+            updatePreview();
+      }
```
This brace should be indented one more level.

sidewayss replied:
check

mattmcclinch commented on mscore/pagesettings.cpp:
```diff
@@ -270,310 +325,392 @@ void PageSettings::orientationClicked()
+      if (flag) { 
+            evenPageLeftMargin ->setValue(oddPageRightMargin ->value());
+            evenPageRightMargin->setValue(oddPageLeftMargin->value());
+      }
```
This brace should be indented one more level.

sidewayss replied:
check

mattmcclinch commented on mscore/pagesettings.cpp:
```diff
@@ -270,310 +325,392 @@ void PageSettings::orientationClicked()
+      else {
+            evenPageLeftMargin ->setValue(oddPageLeftMargin ->value());
+            evenPageRightMargin->setValue(oddPageRightMargin->value());
+      }
```
This brace should be indented one more level.

sidewayss replied:
check

mattmcclinch commented on mscore/pagesettings.cpp:
```diff
@@ -270,310 +325,392 @@ void PageSettings::orientationClicked()
+                  spinOther = isOdd ? evenPageRightMargin : oddPageRightMargin;
+            else 
+                  spinOther = evenPageLeftMargin; // even left margin is disabled
+      }
```
This brace should be indented one more level.

sidewayss replied:
check

mattmcclinch commented on mscore/prefsdialog.cpp:
```diff
@@ -1100,15 +1113,28 @@ void PreferenceDialog::apply()
+                  MScore::setDefaultStyleForParts(new MStyle(MScore::defaultStyle()));
+            else
+                  MScore::readDefaultStyle(partText);
+      }
```
This brace should be indented one more level.

sidewayss replied:
check

mattmcclinch commented on mscore/preferences.cpp:
```diff
@@ -73,8 +73,10 @@ void Preferences::init(bool storeInMemoryOnly)
+	      {PREF_APP_KEYBOARDLAYOUT,                              new StringPreference("US - International")},
+	      {PREF_APP_PAGE_UNITS_GLOBAL,                           new BoolPreference(true, false)},
+	      {PREF_APP_PAGE_UNITS_VALUE,                            new IntPreference(0, false)},
+	      {PREF_APP_PATHS_INSTRUMENTLIST1,                       new StringPreference(":/data/instruments.xml", false)},
```
These four lines have two extra spaces in front.

mattmcclinch replied:
Maybe not. It appeared that way in the diff.

mattmcclinch replied:
Ah, some of the spaces are actually tabs. You should correct that.

sidewayss replied:
That's odd.  They don't appear that way in VS2017.  I'll try opening them in another editor.  They do show up as indented in Notepad.  It's because there's tab character in there at the start of the line.  Fixed.

sidewayss commented:
I am having second thoughts about reverting the file format version number.  The reading of the XML is all done sequentially, not random access.  The QXMLStreamReader class, from which XMLReader inherits, is a stream, sequential.  The file format version is the first thing in the file.  There is one place where I would need to seek in and find the "spatium" tag inside the Style tag, inside the Score tag, in MasterScore::read1():
https://github.com/musescore/MuseScore/blob/15121d7df702fc3582f85fbe80e54a1274fd8064/libmscore/scorefile.cpp#L957

I'd rather not mess around and create some kludgey code here, and the warning and Ignore button encourages people to get the most recent version :).  Does that constitute compatibility, or is the message and Ignore button a deal killer?

sidewayss commented on libmscore/style.cpp:
```diff
@@ -2238,8 +2259,16 @@ void MStyle::load(XmlReader& e)
-                  set(Sid::spatium, e.readDouble() * DPMM);
+            else if (tag == "spatium") { ///!!!3.02 lowercase
+                  set(Sid::spatium, e.readDouble() * DPI_F);
+            }
```
you are correct, thx for noticing.  I just removed the braces locally.

sidewayss commented on libmscore/style.cpp:
```diff
@@ -2345,6 +2376,164 @@ void MStyle::reset(Score* score)
+            set(Sid::marginOddBottom,  value(Sid::pageOddBottomMargin) .toDouble() * PPI);
+            set(Sid::marginEvenTop,    value(Sid::pageEvenTopMargin)   .toDouble() * PPI);
+            set(Sid::marginEvenBottom, value(Sid::pageEvenBottomMargin).toDouble() * PPI);
+      }
```
check

sidewayss commented on libmscore/style.cpp:
```diff
@@ -2345,6 +2376,164 @@ void MStyle::reset(Score* score)
+      if (_isMMInch) { ///!!!
+            psid = QPageSize::id(QSizeF(w, h), QPageSize::Point, QPageSize::FuzzyOrientationMatch);
+            set(Sid::pageSize, int(psid));
+      }
```
check

anatoly-os replied:
It is still the case. Why did you resolve the conversation?

sidewayss commented on mscore/pagesettings.cpp:
```diff
@@ -109,158 +111,211 @@ void PageSettings::setScore(Score* s)
+                  idxType = PAPER_TYPE_METRIC;
+            else
+                  idxType = PAPER_TYPE_IMPERIAL;
+      }            
```
check

sidewayss commented on mscore/editstyle.cpp:
```diff
@@ -1447,5 +1453,53 @@ void EditStyle::resetUserStyleName()
+      preferences.setPreference(isParts ? PREF_SCORE_STYLE_PARTSTYLEFILE
+            : PREF_SCORE_STYLE_DEFAULTSTYLEFILE,
+            filePath);
+}
```
check

mattmcclinch commented on mscore/pagesettings.cpp:
```diff
@@ -41,38 +41,40 @@ PageSettings::PageSettings(QWidget* parent)
+      connect(pageWidth,            SIGNAL(valueChanged(double)), SLOT(widthChanged(double)));
+      connect(pageHeight,           SIGNAL(valueChanged(double)), SLOT(heightChanged(double)));
+      connect(spatiumEntry,         SIGNAL(valueChanged(double)), SLOT(spatiumChanged(double)));
+	connect(oddPageTopMargin,     SIGNAL(valueChanged(double)), SLOT(otmChanged(double)));
```
There is a tab at the beginning of this line instead of 6 spaces.

sidewayss replied:
fixed - how does that even happen?  I have never had the insert tabs option set.

sidewayss commented:
That's my opinion too, but I'm not running the show.  I still need to test further opening the 3.02 files with the current master.  I'll finish that up in the morning and commit the changes.  Then I'll re-configure mingw and try to build the mtests.

mattmcclinch commented:
Why do you need to call `style().setMMInch(true);` in `MasterScore::read1()`? Can't you do it if and when the `<spatium>` tag is read?

sidewayss commented:
I was looking at that last night.  Oops, I have the master checked out right now.  Yes, it's because of read206.cpp.  Though I could potentially make a change inside that module to compensate.  read206 does its own xml parsing, so it does look like I could manage it that way.
I'm in the middle of testing the master with the new format.  I'll give that a try when I'm done.

sidewayss commented:
Turns out that reverting to 3.01 was far simpler than I anticipated.  It's because of the way I isolated 3.01 from read206 and read114 originally.  I was anticipating this being a non-compatible situation and that the entire _isMMInch construct was temporary.  _isMMInch only applies to 301 because I've written extra code in read206.cpp to handle all the style entries there anyway.  I have not touched read114.cpp, and so far that it working fine in my tests; but it never calls MStyle::toPageLayout() during the read114 load, so I might need to double-check that things are OK there.

I'm going to test some more and try to run the mtests before I commit again.  It looks like this won't make it into the upcoming 3.0.1 release (not to be confused with file format version 3.01 or 301), so I'm not going to hurry :)

sidewayss commented:
It passed!
Now the real review can begin.  I have various comments/questions embedded in the code using "///!!!" to begin each comment.  Some of those belong in the review process, and I'll start marking that up.

sidewayss commented on libmscore/style.h:
```diff
@@ -1101,17 +1187,31 @@ class MStyle {
+      MPageLayout _pageOdd;
+      MPageLayout _pageEven;
+      bool _isMMInch;       ///!!!for 3.01 to 3.01+ conversion
+      QMarginsF _marginsOdd; ///!!!Travis workaround
```
These are part of a more refined workaround that did not work on Travis.  I will remove them from here and style.cpp, where they are fully commented out.

sidewayss commented:
The one thing that I have not touched is the MusicXML code in importxmlpass1.cpp and exportxml.cpp.  That code still works, but it still uses the older styles.
I'll take a deeper look at those now.  They use an unusual form of units that I need to reassess.

sidewayss commented:
That last failure is due to Travis taking too much time, too high a server load maybe.  No actual error in the code.  The next rebase/push should turn everything green again.

sidewayss commented:
The Music XML changes in exportxml and importmxmlpass1.cpp are complete.

sidewayss commented:
The current Travis failure is due to a couple of mtest changes in the master that I have not reconciled yet.  I'll do that soon, especially if there is a deadline for merging.
Something worth noting: I had to create a second row for buttons at the bottom of the edit style dialog.  I have attached a screenshot with all the buttons visible.  Maybe the Save As and Reset buttons should be in a menu, or a drop-down of some kind, but this works for now and allows for longer/wider translations of the button text
![editstylebutts](https://user-images.githubusercontent.com/15235202/53648128-c2f6ca00-3c04-11e9-8250-b91cfc8c71c3.png)
.

sidewayss commented:
I just made another change to the Page Settings dialog (pagesettings.ui).  I removed the Apply button, so that it is the same as the Edit Style dialog (editstyle.ui).  Apply is problematic because of the way `undoChangeStyleVal()` checks to see if the value has changed.  It creates a bug where clicking Apply, then clicking OK causes an undo of the Apply.
editstyle.cpp makes changes immediately, but waits until the OK buttons is pressed before executing `endCmd()`.  pagesettings.cpp waits until Apply or OK to make any changes.  All of this could be changed, but it would require significant reorganization of the code, and I am leaving that to a future PR.  I hope that removing this Apply button is not a serious problem for anyone.

sidewayss commented:
I have encountered a problem with making pagesettings operate the same way as editstyle: my new Apply to All Parts button has the same issue as the basic Apply button.  If there are no changes to the styles, it undoes the last style change for that score. I will be leaving an Apply button on the page settings dialog, and adding a feature to check for changes, and `isDirty()` function.  That will solve all of this.

sidewayss commented:
Undo is now working correctly for pagesettings.  My previous comments on the subject are of mixed accuracy.  There is now an isDirty() kind of `if` block at the top of `PageSettings::applyToScore()`: https://github.com/musescore/MuseScore/blob/b97b0cc106069f9021e5ba9a1f8f807ce2eb79f7/mscore/pagesettings.cpp#L661
It prevents phantom undo actions when the user presses Apply or OK without any changes having been made.
The Apply to All Parts button now treats the set of parts as one undo action, the same as the editstyle dialog.
The page settings dialog has a bunch of issues.  I think I fixed them all in this PR.  We'll see...

sidewayss commented:
The last push was to back out my tempo inspector change that duplicates another PR #4253.

anatoly-os commented:
This PR looks untested. I played few minutes and already found several bugs: 
https://youtu.be/_-OtwODLRTA.

Looking through the code...

* I see mentioning of 3.02 here and there. Why? You didn't change the fileformat version
* `spatium` tag in lowercase doesn't look good if we already use `Spatium` tag. Why didn't you find proper name?

sidewayss commented:
Comments fixed.
Landscape bug for custom page sizes fixed.
Horizontal line (separator) added to page settings dialog.

sidewayss commented:
Regarding the `spatium` tag.  It will go away in v4 MuseScore, and it matches `Sid::spatium`.  It's a good long term tag name.

sidewayss commented:
Dmitri - I missed one of your comments - github makes a mess of the screen trying to review them all.  And now I cannot find it after my latest push.
You are correct that those hard-coded millimeter spatium values are in at least 2 places, I thought it was 3.  Originally I thought the code would be obsolete except for pre 3.01 scores, but now that is not the case.  I can make it a separate function if you like.
Do you want me to do that in this PR?

sidewayss commented:
Dmitri - Travis is done, and so am I.  I have addressed all of your comments, AFAICT.  I have reviewed the code for purely style/indent changes and reverted what I found, so it should be easier to read the diffs.  I'm about to commit, squash, and push again now.
The one outstanding issue is whether you want to move that tablature tie fix to a separate PR.

sidewayss commented:
github is accepting my edits to the previous comment, but not saving them.
I have fixed two more spots with style/indent issues and pushed again.

sidewayss commented:
Oh no.  My changes to exportxml.cpp and importmxmlpass1.cpp are gone.  Overwritten in a rebase?  I have no idea.  Now my commits are all squashed and the code is gone.  My last non-github backup is from before these changes, so I have to rewrite this now.  Not too complicated.  I remember it more or less.  It would be nice to know what happened.  I don't want it to happen ever again.  The current code works, but it should be updated to not use the older page styles.

dmitrio95 commented:
> Oh no. My changes to exportxml.cpp and importmxmlpass1.cpp are gone. Overwritten in a rebase?

Try `git reflog mscore/exportxml.cpp`, maybe it will help you find the lost changes.

> Should I resolve these conversations, or is that for you to do? This is my first PR to ever be reviewed this way. I'll commit, squash, and push now.

I cannot mark them resolved myself as I am not the pull request author so if you prefer them collapsed you should probably mark them resolved yourself.

> The only outstanding issues are whether you want to move that tablature tie fix to a separate PR, and whether you want to move locale to mtest.

Yes, if this is not related to the main topic of this pull request I believe it would be better to have it in a separate PR. Then both changes can be reviewed independently.

sidewayss commented:
I'm looking at `git reflog` results, and there are the commits.  How do I get to that code?  I'll look for docs online...

sidewayss commented:
Dmitri - If you're interested in changes to unit conversions, you can now examine mscore/exportxml.cpp.  It doesn't look like the disappearance is related to a rebase, but I'm still not certain.  I copied the older commit's two files, (importmxmlpass1.cpp too), rebased to the latest master, built it, tested a bit, and just now pushed it.
Here is my Excel spreadsheet, in case you want to dig in the the math in detail - it lays out the 3.0 formulas using the default spatium value of 25 - it might be too undocumented to be helpful:
[tenths.xlsx](https://github.com/musescore/MuseScore/files/2975278/tenths.xlsx)

I resolved all the review comments that were completely resolved.  I left open anything with a question or discussion still possibly unresolved.

sidewayss commented:
I have created a new PR for the tablature tie issue: #4804, and I have removed the code from this PR's tie.cpp.  Now only 538 files changed :)

sidewayss commented:
Another (hopefully final) push to remove some style-only changes in exportxml.cpp.  The only outstanding issues are:
1) locale maybe moves to mtest
2) concerns about storing page settings in Points x 10,000 units
Anything else?

sidewayss commented:
I have updated the PDF file I wrote originally to explain and justify these changes.  It's less justification and more explanation now, and the screenshots are updated.
[MS_PageSettings.pdf](https://github.com/musescore/MuseScore/files/2975713/MS_PageSettings.pdf)

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2140,7 +2158,29 @@ void MStyle::precomputeValues()
+      QVariant val  = value(idx);
+      QVariant base = MScore::baseStyle().value(idx);
+      const char* type = styleTypes[int(idx)].valueType();
+      if      (!strcmp(type, "bool"))
```
Spaces.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2140,7 +2158,29 @@ void MStyle::precomputeValues()
+      QVariant base = MScore::baseStyle().value(idx);
+      const char* type = styleTypes[int(idx)].valueType();
+      if      (!strcmp(type, "bool"))
+            return val.toBool()   == base.toBool();
```
Spaces. There is no need to visually align the cases.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2140,7 +2158,29 @@ void MStyle::precomputeValues()
+      else if (!strcmp(type, "Ms::Align"))
+            return Align(val.toInt()) == Align(base.toInt());
+      else if (!strcmp(type, "QColor"))
+            return val.value<QColor>()        == base.value<QColor>();
```
Spaces.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2140,7 +2158,29 @@ void MStyle::precomputeValues()
+      else if (!strcmp(type, "Ms::Spatium"))
+            return val.value<Spatium>().val() == base.value<Spatium>().val();
+      else
+            return val == base; // never happens: rejected by MStyle::read()
```
If it never happens, add `Q_ASSERT` here.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2140,7 +2158,29 @@ void MStyle::precomputeValues()
 
 bool MStyle::isDefault(Sid idx) const
       {
-      return value(idx) == MScore::baseStyle().value(idx);
```
Why? Does it fix an existing bug? If it is, add the changes as a separate commit with fix #XXXXXX mention.

anatoly-os replied:
@sidewayss please address the comment.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2361,11 +2403,21 @@ bool MStyle::load(QFile* qf)
+void MStyle::spatium301(XmlReader& e) {
+      _isMMInch = true;
+      QString txt = e.readElementText();
+      if (txt == "1.76389" || txt == "1.764" || txt == "1.7526" || txt == "1.753")
```
What do these values mean? Imagine a developer who wants to edit this code or fix some bug, how will he know the purpose and the meaning of the numbers?

anatoly-os replied:
@sidewayss please address the comment.

anatoly-os replied:
Ok, now the purpose of the method is clear, but the values are still better to be stored as `static constexpr` variables in the method. If you know why the values are different, you can mention it in variable names. If not, link the place in code (from the history maybe) which specifies the hardcoded values. The reason is not to forget the genesis of the values.

sidewayss replied:
These are string values.  Are you suggesting that I declare them as `static constexpr QString`?  I cannot do it with char*.  And why should they be static inside the method?

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2381,8 +2431,12 @@ void MStyle::load(XmlReader& e)
                   }
-            else if (tag == "Spatium")
-                  set(Sid::spatium, e.readDouble() * DPMM);
+            else if (tag == "spatium") { // 3.01+ lowercase
```
I don't like the same names which are different by capitalisation. Better name is needed. "SpatiumNew", "UniversalSpatium"?

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2526,217 @@ void MStyle::save(XmlWriter& xml, bool optimize)
             }
-      xml.tag("Spatium", value(Sid::spatium).toDouble() / DPMM);
+      xml.tag("Spatium", value(Sid::spatium).toDouble() / DPMM);                 // 3.01
+      xml.tag("spatium", rint(value(Sid::spatium).toDouble() / DPI_F * MSCX_F)); // 3.01+ last one wins at load time
```
The comment specifies some unreliable behaviour. What does "wins" mean?

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2526,217 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+                                          : QPageSize::Letter;
+            }
+      _pageSize = QPageSize(psid);           
+      _pageOdd  = MPageLayout();
```
Please, don't use extra spaces to visually align the variables. It will be a pain in ass for anyone who uses auto-formatter.

anatoly-os replied:
@sidewayss please address the comment.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2526,217 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+            QRectF    rect = _pageOdd.fullRect(QPageLayout::Inch);
+            QMarginsF marg = _pageOdd.margins (QPageLayout::Inch);
+
+            if (abs(value(Sid::pageWidth) .toDouble() - rect.width())  >= 0.01)
```
Extract the precision value as separate variable. Extract a method which compares two values with specified precision.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2526,217 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+//------------------------------------------------------------------------------
+void MStyle::fromPageLayout()
+      {
+      if (!MScore::testMode) { ///!!!Travis workaround
```
What do you mean?

anatoly-os replied:
@sidewayss please address the comment.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2526,217 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+
+      qs      = value(Sid::pageUnits).toString();
+      sz      = sizeof(pageUnits) / sizeof(pageUnits[0]);
+      for (int i = 0; i < sz; i++) {  // units stored as string for readability
```
Why not to store them as Enum? We won't need more than Imperial/Metric units.

anatoly-os commented on libmscore/style.h:
```diff
@@ -1094,6 +1106,80 @@ enum class Sid {
+      qreal       _factor;      // conversion-to-points factor
+
+public:
+      const char*  key()         const { return _key;             }
```
Shouldn't all these methods be `const`?

anatoly-os commented on libmscore/style.h:
```diff
@@ -1094,6 +1106,80 @@ enum class Sid {
+      double factor()      { return pageUnits[int(units())].factor();      }
+      double paintFactor() { return pageUnits[int(units())].paintFactor(); }
+   public:
+      double width()        { return fullRect().width()  * paintFactor(); }
```
`const` methods are the only solution here. Described above.

anatoly-os commented on libmscore/style.h:
```diff
@@ -1129,7 +1227,11 @@ class MStyle {
 
-      void reset(Score*);
+      void fromPageLayout();
+      void   toPageLayout();
```
Spaces.

anatoly-os commented on mscore/exportxml.cpp:
```diff
@@ -265,6 +265,8 @@ class GlissandoHandler {
 typedef QHash<const Chord*, const Trill*> TrillHash;
 typedef QMap<const Instrument*, int> MxmlInstrumentMap;
 
+#define TEN 10 // there's a reason for the name "tenths", must be global for static funcs
```
Why? Why not to use constexpr static method? Same for TENTHS.

anatoly-os commented on mscore/musescore.cpp:
```diff
@@ -7368,10 +7365,19 @@ int main(int argc, char* av[])
+                  def .setPageSize(ps);
+                  odd .setPageSize(ps);
+                  even.setPageSize(ps);
+// minimum margins is a nice idea, but it implies additional code elsewhere
```
What do you mean? Why is the following code needed?

anatoly-os commented on mscore/pagesettings.cpp:
```diff
@@ -108,159 +105,213 @@ void PageSettings::setScore(Score* s)
 //---------------------------------------------------------
-//   updateValues
-//    set gui values from style
+//   updateWidgets
```
there is no need to rename the method. Why? It doesn't update widgets, it updates values in the controls.

anatoly-os replied:
@sidewayss please address the comment.

sidewayss replied:
And it updates much more than the values.  It updates all the configured settings for each widget.

anatoly-os commented on mscore/pagesettings.cpp:
```diff
@@ -270,311 +321,428 @@ void PageSettings::orientationClicked()
+            {
+            MPageLayout& sOdd  = score->style().pageOdd();
+            MPageLayout& sEven = score->style().pageEven();
+            if (sOdd.units()               == odd.units()
```
Please put each `&&` on the previous strings according to the code style.

anatoly-os commented on mscore/pagesettings.cpp:
```diff
@@ -270,311 +321,428 @@ void PageSettings::orientationClicked()
+      score->undoChangeStyleVal(Sid::marginEvenTop,    even.topMarginPoints());
+      score->undoChangeStyleVal(Sid::marginEvenBottom, even.bottomMarginPoints());
+
+      // older 3.01 styles
```
Older? There is one 3.01 file format. A developer won't understand this comment.

anatoly-os commented on mscore/pagesettings.cpp:
```diff
@@ -270,311 +321,428 @@ void PageSettings::orientationClicked()
-            case 1:
-                  preview->score()->doLayout();
+            case QDialogButtonBox::Ok:
+                  applyToScore(cs); // FALLTHROUGH
```
There is no big deal to add two extra strings, but solves possible issues for future development because of fallthrough cases.

anatoly-os commented on mscore/pagesettings.cpp:
```diff
@@ -270,311 +321,428 @@ void PageSettings::orientationClicked()
+
+void PageSettings::done(int val)
+      {
+      cs->setLayoutAll();     // HACK
```
HACK? Is it possible to avoid any hacks?

anatoly-os replied:
@sidewayss could you please address (answer to) the comment?

anatoly-os commented on mscore/pagesettings.h:
```diff
@@ -22,38 +22,47 @@ class MasterScore;
 class Score;
 class Navigator;
 
+#define PAPER_TYPE_METRIC   0
```
Why not to use enums?

anatoly-os commented on mtest/biab/updateReference:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
New file? Why do we need it in the repo?

anatoly-os replied:
@sidewayss could you please address (answer to) the comment?

sidewayss commented:
I didn't wait for travis or appveyor to complete prior to my last set of pushes.  Now the PR seems to be running only appveyor, not travis.  Please let me know if there is something I should do, or if I should just wait for my next push to resolve this.

anatoly-os commented:
@sidewayss please comment or resolve the review comments along with the related changes.

Also, "This is not intended to be a final PR for the master" is not true, isn't it?

sidewayss commented:
Yes, this is now intended to be the final PR.
I have commented or resolved all the review comments.

sidewayss commented:
Today's push includes the removal of more non-compliant spacing, especially in variable declarations, which is what I believe @anatoly-os was pointing out in one of his comments.  I'm hoping that any and all style issues are resolved now.

sidewayss commented:
Today's push is a rebase that required the updating of mtest files for the new test: tst_unrollrepeats.  I added an updateReference script to that directory for future -ref.mscx file maintenance.

Jojo-Schmitz commented:
You may want to change the commit title to 
Fix #280363 page settings and units changes
i.e. add a a space before the #

sidewayss commented:
Yes, I was going to do that, then I waited a couple of days before committing again.  I'll remember with the next commit by fixing the message now.  Thanks for the reminder.

sidewayss commented:
I have added some stdout messages that will illustrate the Travis malfunction.  ctest logs the output to MuseScore/build.debug/mtest/Testing/Temporary/LastTest.log.
I was unable to restart my own repo's most recent Travis job in debug mode.  Hopefully that will be resolved early tomorrow.  I sent an email to Travis support.  Once I have access to the Travis version of that file I'll post it and my local version here so you can see what's happening.

sidewayss commented:
Regarding the idea of setting the locale in mtest: I don't see a MuseScore command line option to do that.  mtest runs MuseScore via the command line, correct?
I could add a new command line option, but that would be the same as using `MScore::testMode`.

But there are two cases where I use `if (MScore::testMode)`, and I can consolidate those into one case in `Preferences::init()`, where I currently set the base units based on locale:
https://github.com/sidewayss/MuseScore/blob/840774743f9645bf4f765645aceee13977becd71/mscore/preferences.cpp#L72
In the end it's not the locale, it's the units that must be set to Millimeters (I use millimeters because the base style is defined in mm, and mm are the traditional MuseScore default user units).

I have made this change, tested, and pushed.  The link above points to the new code.

sidewayss commented:
Regarding the name of the new XML tag for spatium: This is a unique case where there is a new tag, but it shares the Sid with the old tag.  If the new tag is in the file, then it overrides the old tag's value for `Sid::spatium`.  Adding an additional Sid for spatium seemed like a bad idea to me, but it is certainly an option.  It would change code in several other places that now rely on `Sid::spatium`, but there are not many of those.  What do you all think about this?

I can rename the tag "spatiumNew", or "spatium2019", or "staffSpace", or "spatiumPt10K" to indicate that it is stored in Points x 10,000.  I'm open to other name suggestions as well.  I'd like to settle on an acceptable name before making the change, as it requires the regeneration of all the mtest -ref files.  I like the the tag as I have it now, but if it's not acceptable to others, then I'll change it to something that is.

sidewayss commented:
Attached are two log files.  One is from my last local run of ctest, the other is from my last Travis build.  The Travis log has zero values all over the place for the odd left margin.  If you look at style.cpp, in `initPageLayout()`, `toPageLayout()`, and `fromPageLayout()`, you'll see the `cout` calls that generate the "Odd Left" log entries.  The source/origin values are always valid, default values of 10mm (0.39in),  but those values often fail to assign to the QPageLayout objects when running in Travis.  The result is margin values of zero.
For example: in the Travis log, the top of each test has two entries for initPageLayout, and many of them look like this:
`initPageLayout - Odd Left Margin: 0`
`fromPageLayout - Odd Left Margin: 0`
`initPageLayout - Odd Left Margin: 0.39`
`fromPageLayout - Odd Left Margin: 0.39`

The first two lines are for the base style, the second two lines are for the default style.  It makes no sense that one would work and the other would fail.  These results in the Travis log are not 100% consistent.  Some of the tests have fewer zero margin values than others

[LastTest.log](https://github.com/musescore/MuseScore/files/3006051/LastTest.log)
[travislog.txt](https://github.com/musescore/MuseScore/files/3006052/travislog.txt)

anatoly-os commented on libmscore/cmd.cpp:
```diff
@@ -3600,7 +3591,6 @@ void Score::cmd(const QAction* a, EditData& ed)
             { "pad-note-64-TAB",            [this]{ padToggle(Pad::NOTE64);                                     }},
             { "pad-note-128",               [this]{ padToggle(Pad::NOTE128);                                    }},
             { "pad-note-128-TAB",           [this]{ padToggle(Pad::NOTE128);                                    }},
-            { "reset-style",                [this]{ cmdResetStyle();                                            }},
```
@sidewayss please address the comment.

anatoly-os commented on libmscore/score.cpp:
```diff
@@ -275,22 +275,14 @@ Score::Score(MasterScore* parent, bool forcePartStyle /* = true */)
-                  Sid::spatium
-                  };
-            // but borrow defaultStyle page layout settings
-            for (auto i : styles)
```
@sidewayss please address the comment.

anatoly-os commented on libmscore/score.h:
```diff
@@ -459,7 +459,6 @@ class Score : public QObject, public ScoreElement {
       ChordRest* nextMeasure(ChordRest* element, bool selectBehavior = false, bool mmRest = false);
       ChordRest* prevMeasure(ChordRest* element, bool mmRest = false);
       void cmdSetBeamMode(Beam::Mode);
-      void cmdResetStyle();
```
@sidewayss please address the comment.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2381,8 +2433,12 @@ void MStyle::load(XmlReader& e)
-            else if (tag == "Spatium")
-                  set(Sid::spatium, e.readDouble() * DPMM);
+            else if (tag == "spatium") { ///!!!3.01+ lowercase
+                  set(Sid::spatium, e.readInt() / MSCX_F * DPI_F);            
```
@sidewayss please address the comment.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2534,221 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+//------------------------------------------------------------------------------
+void MStyle::fromPageLayout()
+      {
+cout << "fromPageLayout - Odd Left Margin: " << _pageOdd.margins(QPageLayout::Inch).left() << endl;
```
Use `qDebug()` instead of `cout` to avoid excessive output in release versions... And spacing is wrong.

sidewayss replied:
As explained in a comment on the PR page, these are temporary and purely for the purpose of illustrating the Travis issue that I can only work around.  The spacing is on purpose to call them out as temporary.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2531,217 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+      // 3.01+ styles
+      set(Sid::pageFullWidth,  _pageOdd.widthPoints());
+      set(Sid::pageFullHeight, _pageOdd.heightPoints());
+      if (!MScore::testMode) { ///!!!Travis workaround
```
@sidewayss please address the comment.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2534,221 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+      _pageOdd.setPageSize(_pageSize);
+      _pageOdd.setUnits(QPageLayout::Unit(idxUnit));
+      _pageOdd.setMargins(oddMarg);
+cout << "  toPageLayout - Odd Left Source: " <<  oddMarg.left() << endl;
```
See above

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2534,221 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+      _pageOdd.setUnits(QPageLayout::Unit(idxUnit));
+      _pageOdd.setMargins(oddMarg);
+cout << "  toPageLayout - Odd Left Source: " <<  oddMarg.left() << endl;
+cout << "  toPageLayout - Odd Left Margin: " << _pageOdd.margins(QPageLayout::Inch).left() << endl;
```
See above

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2534,221 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+      right  = value(Sid::marginOddRight).toInt()  / factor;
+      top    = value(Sid::marginOddTop).toInt()    / factor;
+      bottom = value(Sid::marginOddBottom).toInt() / factor;
+cout << "  toPageLayout - Odd Left Origin: " << left << endl;
```
See above

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2531,217 @@ void MStyle::save(XmlWriter& xml, bool optimize)
+      switch (unit) {
+            case QPageLayout::Millimeter:
+            case QPageLayout::Didot:
+            case QPageLayout::Cicero:
```
@sidewayss please address the comment.

anatoly-os commented on libmscore/style.cpp:
```diff
@@ -2474,18 +2534,221 @@ void MStyle::save(XmlWriter& xml, bool optimize)
       {
-      for (const StyleType& st : styleTypes)
-            score->undo(new ChangeStyleVal(score, st.styleIdx(), MScore::defaultStyle().value(st.styleIdx())));
+      return(abs(value(sid).toDouble() - pageValue) >= 0.01);
```
Please replace `0.01` with `epsilon` variable which should be a parameter of the method with default value `0.01`.

sidewayss replied:
In the next push, later tonight.

anatoly-os commented on mscore/exportxml.cpp:
```diff
@@ -3660,7 +3659,7 @@ void ExportMusicXml::textLine(TextLine const* const tl, int staff, const Fractio
             else
                   lineEnd = "down";
-            rest += QString(" end-length=\"%1\"").arg(hookHeight * 10);
+            rest += QString(" end-length=\"%1\"").arg(hookHeight * TEN);
```
I don't like using TEN constant here and there. I would create a method which encapsulates the meaning of this operation. If the operation relates to musicXmlExport only, let's call it `convertValueToMXLExportSemantic`. The method will multiply values by 10. In this case we won't need the global constant value. Also, if we decide to refactor the code, all the changes will be obvious.

My general point is that the code should be clear. Using TEN instead of 10 doesn't make code clearer. The semantic of the TEN should be highlighted and ideally encapsulated.

sidewayss replied:
I see your point.  I'll make the change.

anatoly-os commented on mscore/importmxmlpass1.cpp:
```diff
@@ -542,7 +542,7 @@ static void doCredits(Score* score, const CreditWordsList& credits, const int pa
       */
       // page width, height and odd top margin in tenths
-      const double ph  = score->styleD(Sid::pageHeight) * 10 * DPI / score->spatium();
+      const double ph = score->styleI(Sid::pageFullHeight) * 10 / MSCX_F / score->spatium();
```
again magic 10

sidewayss replied:
Yes, but it's in a different module :)  I'll have to make the new function public and make the class available here too.

anatoly-os commented on mscore/magbox.cpp:
```diff
@@ -172,12 +172,12 @@ double MagBox::getMag(ScoreView* canvas) const
                   else {
-                        mag1 = cw / (pw * 2 * DPI + 50);
-                        mag2 = ch / (ph * DPI);
+                        mag1 = cw / (pw * 2 + 50);
```
why 50?

sidewayss replied:
I did not change this, so I have no idea.

anatoly-os commented on mscore/pagesettings.cpp:
```diff
@@ -108,159 +105,217 @@ void PageSettings::setScore(Score* s)
-      updatePreview(0);
+void PageSettings::sizeChanged(int idx)
+      {
+      if (idx >= 0) // should never be -1, but just in case...
```
add `Q_ASSERT()` to make sure the value is never negative.

sidewayss replied:
OK, no problem.

anatoly-os commented on mscore/pagesettings.cpp:
```diff
@@ -270,311 +325,430 @@ void PageSettings::orientationClicked()
+      double newVal;
+
+      QPageLayout::Unit unit = score->style().pageOdd().units();
+      if ((unit == QPageLayout::Millimeter && val == 1.764) ||
```
Magic numbers. Please create variables with sensible names.

sidewayss replied:
Even though they are used only once?  A comment describing them is not enough? I'm fine with creating local constants if that is preferred.
I see literal values all over the code, so I assume that this is OK.

sidewayss replied:
I have created variables, and commented them.

anatoly-os commented on mscore/pagesettings.cpp:
```diff
@@ -270,311 +325,430 @@ void PageSettings::orientationClicked()
-      updatePreview(0);
+      QPageLayout::Unit u = QPageLayout::Unit(unitsList->currentIndex());
+      MStyle& style = preview->score()->style();
+      style.pageOdd ().setUnits(u);
```
Spaces

sidewayss replied:
I already fixed this but haven't yet pushed.  I've done a lot of space removing.

anatoly-os commented:
@sidewayss I see you addressed some comments in the PR's discussion. Could you please comment on the exact reviews to keep things focused? I left more comments. Some previous comments are not addressed.

sidewayss commented:
I added comments here in addition to review comments because I was trying to keep the conversation focused on the review comments that are still unresolved.  It's all duplicate information.  AFAIK all previous comments have been addressed via a response comment and/or resolving them.  I'll look at your new comments now and respond.

sidewayss commented:
If you are seeing comments unaddressed that is the fault of github.  I am seeing some comments not displayed here, but they are displayed in the file diff, or the review comment has disappeared.

sidewayss commented:
I'm so sorry.  I did not know that I had to click the green button to submit my review comments.  They showed for me so I assumed they showed up for everyone else.  They are now all submitted.  I have a couple of changes to make in response to @anatoly-os's latest comments.  I'll push those changes later this morning.

sidewayss commented:
The latest push sets the margins/width/height spinboxes' `decimals` property to 0 for Points & Didot and 2 for all other units.  That matches QPageLayout's internal rounding scheme, so the dialog and QPageLayout are now fully in sync.

sidewayss commented on libmscore/style.cpp:
```diff
@@ -2485,6 +2541,251 @@ void MStyle::reset(Score* score)
+//                    must wait until setMscoreLocale() is called in main()
+//--------------------------------------------------------------------------
+
+void MStyle::initPageLayout()
```
Here is the only example of an issue that I can see, and it is limited to these circumstances:
1) Your score uses the default page size
2a) Someone else in a different locale opens your score in their default page size, and it does not look the same. It won't look too different because A4 and Letter are similar sizes, but it might break systems differently, or even create extra pages because of overflow.
2b) You toggle your preference to/from mm/inch in the new preferences dialog. The next time you open a default-page-size score (nothing written to the file) it will be in a different page size.

On the other hand, the reason for not storing the default page size is so that MuseScore templates open in the user's default page size, not always in A4.  That's a very positive feature, and should be maintained.

The alternative is to always store the page size, regardless of whether it matches the default.  But that means either templates will all be A4, or there must be a special "save as template" feature that does not store the page size.

MarcSabatella replied:
That's a real concern, though.  Lett and A4 are different enough that it can force measures to different systems & systems to different pages.

We do have special handling to make sure newly created scores are either A4 or Letter depending on locale (actually default printer default paper size, I think).  But when opening an existing score as opposed to creating a new one, it's important to honor the page size.  We do have a template save mode, I think it's a command line option.  In the past we've also just special-cased A4 & Letter when creating from templates, though.

sidewayss commented:
New PR created here: https://github.com/musescore/MuseScore/pull/4906

That is the first in a series of PRs that break this PR into pieces.

Jojo-Schmitz commented:
rebase needed

shoogle commented:
@sidewayss 

> New PR created here: #4906
>
> That is the first in a series of PRs that break this PR into pieces.

Good plan, as this one will be a pain to review and a pain to rebase otherwise. However, I recommend starting with the smallest possible changes: one PR for each of them.

For example, you move the reset style button from the [menu](https://github.com/musescore/MuseScore/pull/4456/files#diff-44ed85c6927e1199227c7488016b668fL3624) to the [dialog](https://github.com/musescore/MuseScore/pull/4456/files#diff-2c8141e7db0795529789a8ae53502f4fR83). That doesn't affect anything else, so it can be one PR all on its own.

You also fix [this bug](https://musescore.org/en/node/281445) with the scaling control in page settings. That could be in a PR of it's own, possibly with other changes to the dialog (but only if they are minor otherwise it just means the fix is delayed).

sidewayss commented:
@shoogle - My priority was fixing the spatium rounding and adding additional units to the PageSettings dialog, so that's what #4906 is.  The minor bugs were thrown into this EPIC pull request because they were related and easy to fix.  So they are the lowest priority for me.

I have been away from MuseScore lately working on the other end of my project, but there seems to be some interest around this PR again, so I'll rebase #4906 soon and start reminding the powers that be to review it again.

As for moving the button from the menu to the dialog, that is one that I might have to give up on just because of the lack of space in the dialog.  I had to add a second row of buttons to do it, and that doesn't seem like a good idea anymore, after the other (now merged) fixes I made to the dialog's sizing functionality.

shoogle commented:
> My priority was fixing the spatium rounding and adding additional units to the PageSettings dialog

I understand. Would you mind if I (or anybody else) submits your fixes for the other bugs?

> there seems to be some interest around this PR again

I think everybody wants to see this PR merged (including, I believe, "the powers that be") but this PR introduces lots of new features (and new complexity), so it may be a little while yet before it is merged.

sidewayss commented:
> I understand. Would you mind if I (or anybody else) submits your fixes for the other bugs?

Please do so.  I have no personal stake in taking credit for any of this.

> I think everybody wants to see this PR merged (including, I believe, "the powers that be") but this PR introduces lots of new features (and new complexity), so it may be a little while yet before it is merged.

The complexity and additional features are greatly reduced in #4906.  Hopefully I can get that PR back on track soon.  Thanks.

sidewayss commented:
@shoogle - If you're talking about fixing issues inside pagesettings.ui, please keep me informed.  That module is a tangle in terms of fixing one thing but not the other.  I approached it as a rewrite, not as a set of individual fixes.

dmitrio95 commented:
Closing and archiving this pull request since this is too large for a review.
